src/app/content/utils/seoUtils.ts

```diff
@@ -9,20 +12,82 @@ import {
   splitTitleParts,
 } from './archiveTreeUtils';
 
-const getParentPrefix = (node: LinkedArchiveTreeNode | undefined): string => {
+const domParser = new DOMParser();
+
+export const getParentPrefix = (node: LinkedArchiveTreeNode | undefined, includeTitle: boolean = false): string => {
   if (!node) {
     return '';
   }
 
   if (archiveTreeSectionIsChapter(node)) {
-    const number = getArchiveTreeSectionNumber(node);
-    return `Ch. ${number} `;
+    const number = getArchiveTreeSectionNumber(node).trim();
+    const name = getArchiveTreeSectionTitle(node);
+    return includeTitle ? `Ch. ${number}. ${name}` : `Ch. ${number} `;
   }
 
   return archiveTreeSectionIsBook(node.parent)
     ? getArchiveTreeSectionTitle(node) + ' '
-    : getParentPrefix(node.parent);
+    : getParentPrefix(node.parent, includeTitle);
+
+};
+
+const hideMath = (node: Element) => {
+  node.querySelectorAll('math, .os-math-in-para').forEach((el: Element) => {
+    el.outerHTML = '...';
+  });
+  return node;
+};
+
+export const getTextContent = (str: string) => {
+  const parsed = domParser.parseFromString(str, 'text/html');
+  const text = parsed.body.textContent;
+  return text;
+};
+
+const removeExcludedContent = (node: Element) => {
+  if (!node) {
+    return null;
+  }
+  const excludedContent = node.querySelectorAll(
+    '[data-type="abstract"], .learning-objectives, .chapter-objectives, .be-prepared, .os-teacher'
+  ) || [];
+
+  for (const item of Array.from(excludedContent)) {
+    if (item) {
+      item.remove();
+    }
+  }
+};
+
+export const generateExcerpt = (str: string) => {
+  return str.replace(/\n+/g, ' ').replace(/\s+/g, ' ').trim().substring(0, 152) + '...';
+};
+
+const getPageDescriptionFromContent = (page: HTMLElement): string | null => {
+  if (!page) {
+    return null;
+  }
+  removeExcludedContent(page);
+  // tslint:disable-next-line: max-line-length
+  const paragraphs = [...Array.from(page.querySelectorAll(':scope>section>p')), ...Array.from(page.querySelectorAll(':scope>p'))];
```

Comment:
@TomWoodward I had to revert to using `:scope` because the `div:first-child>p` selector was finding direct-descendent paragraphs of _all_ first-child divs on the page. But based on what you said about page content IDs I think this is ok.

Reply:
I lied, my metadata dump script complains about invalid IDs when I run it : (